File: src/parser.py
import json

def parseDSL(text: str):
    lines = text.split('\n')
    dsl = {'state_schema': {}, 'invariants': [], 'actions': {}}
    current_section = ''
    current_action = ''
    sub_section = ''

    for line in lines:
        line = line.split('#')[0].rstrip()
        if not line:
            continue

        stripped = line.strip()
        indent = len(line) - len(line.lstrip())

        if line.startswith('state_schema:'):
            current_section = 'state_schema'
        elif line.startswith('invariants:'):
            current_section = 'invariants'
        elif line.startswith('action '):
            current_section = 'action'
            current_action = line[7:].replace(':', '').strip()
            dsl['actions'][current_action] = {'requires': [], 'mutates': []}
        elif current_section == 'state_schema' and indent > 0:
            parts = stripped.split(':')
            if len(parts) == 2:
                dsl['state_schema'][parts[0].strip()] = parts[1].strip()
        elif current_section == 'invariants' and indent > 0:
            if stripped.startswith('- NEVER:'):
                dsl['invariants'].append({'type': 'NEVER', 'condition': stripped[8:].strip()})
            elif stripped.startswith('- ALWAYS:'):
                dsl['invariants'].append({'type': 'ALWAYS', 'condition': stripped[9:].strip()})
        elif current_section == 'action' and indent > 0:
            if stripped == 'requires:':
                sub_section = 'requires'
            elif stripped == 'mutates:':
                sub_section = 'mutates'
            elif stripped.startswith('- '):
                dsl['actions'][current_action][sub_section].append(stripped[2:].strip())

    return dsl

File: src/parser.ts
export function parseDSL(text: string) {
  const lines = text.split('\n');
  const dsl: any = { state_schema: {}, invariants: [], actions: {} };
  let currentSection = '';
  let currentAction = '';
  let subSection = '';

  for (let line of lines) {
    line = line.split('#')[0].trimEnd();
    if (!line) continue;

    const stripped = line.trim();
    const indent = line.length - line.trimStart().length;

    if (line.startsWith('state_schema:')) {
      currentSection = 'state_schema';
    } else if (line.startsWith('invariants:')) {
      currentSection = 'invariants';
    } else if (line.startsWith('action ')) {
      currentSection = 'action';
      currentAction = line.substring(7).replace(':', '').trim();
      dsl.actions[currentAction] = { requires: [], mutates: [] };
    } else if (currentSection === 'state_schema' && indent > 0) {
      const parts = stripped.split(':');
      if (parts.length === 2) {
        dsl.state_schema[parts[0].trim()] = parts[1].trim();
      }
    } else if (currentSection === 'invariants' && indent > 0) {
      if (stripped.startsWith('- NEVER:')) {
        dsl.invariants.push({ type: 'NEVER', condition: stripped.substring(8).trim() });
      } else if (stripped.startsWith('- ALWAYS:')) {
        dsl.invariants.push({ type: 'ALWAYS', condition: stripped.substring(9).trim() });
      }
    } else if (currentSection === 'action' && indent > 0) {
      if (stripped === 'requires:') {
        subSection = 'requires';
      } else if (stripped === 'mutates:') {
        subSection = 'mutates';
      } else if (stripped.startsWith('- ')) {
        dsl.actions[currentAction][subSection].push(stripped.substring(2).trim());
      }
    }
  }
  return dsl;
}
